dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/QueryItemHelper.java

```diff
@@ -117,4 +132,98 @@ else if ( item.hasOptionSet() && (option = item.getOptionSet().getOptionByCode(
             return value + itemValue;
         }
     }
+
+    /**
+     * Returns a map of metadata item options and {@link Option}.
+     *
+     * @param grid the Grid instance.
+     * @param params the EventQueryParams.
+     * @return a list of options.
+     */
+    public static List<Option> getItemOptions( final Grid grid, final EventQueryParams params )
+    {
+        final List<Option> options = new ArrayList<>();
+
+        for ( int i = 0; i < grid.getHeaders().size(); ++i )
+        {
+            final GridHeader gridHeader = grid.getHeaders().get( i );
+
+            if ( gridHeader.hasOptionSet() && isNotEmpty( grid.getRows() ) )
+            {
+                options.addAll( getItemOptionsThatMatchesRows( grid, i, gridHeader ) );
+            }
+            else if ( gridHeader.hasOptionSet() && isEmpty( grid.getRows() ) )
+            {
+                options.addAll( getItemOptionsWhenRowsIsEmpty( params ) );
+            }
+        }
+
+        return options.stream().distinct().collect( toList() );
+    }
+
+    private static List<Option> getItemOptionsWhenRowsIsEmpty( final EventQueryParams params )
+    {
+        final List<Option> options = new ArrayList<>();
+
+        if ( isNotEmpty( params.getItems() ) )
+        {
+            final List<QueryItem> items = params.getItems();
+
+            for ( final QueryItem item : items )
+            {
+                final boolean hasOptions = item.getOptionSet() != null
+                    && isNotEmpty( item.getOptionSet().getOptions() );
+
+                if ( hasOptions && isNotEmpty( item.getFilters() ) )
+                {
+                    options.addAll( getItemOptionsBasedOnTheFilter( item ) );
+                }
+            }
+        }
+
+        return options;
+    }
+
+    private static List<Option> getItemOptionsThatMatchesRows( final Grid grid, final int columnIndex,
+        final GridHeader gridHeader )
+    {
+        final List<Option> options = new ArrayList<>();
+
+        options.addAll( gridHeader
+            .getOptionSetObject()
+            .getOptions()
+            .stream()
+            .filter( opt -> opt != null && grid.getRows().stream().anyMatch( r -> {
+                Object o = r.get( columnIndex );
+                if ( o instanceof String )
+                {
+                    return ((String) o).equalsIgnoreCase( opt.getCode() );
+                }
+
+                return false;
+            } ) ).collect( toList() ) );
+
+        return options;
+    }
+
```

Comment:
Add Javadoc to explain what is going on here.

Reply:
Done